esi/lib/node.py

```diff
@@ -0,0 +1,108 @@
+
+    # base network information
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        f1 = executor.submit(connection.network.ips)
+        f2 = executor.submit(connection.list_networks)
+        f3 = executor.submit(
+            utils.get_ports, connection, filter_network)
+        floating_ips = list(f1.result())
+        networks = f2.result()
+        networks_dict = {n.id: n for n in networks}
+        neutron_ports = f3.result()
+
+    # update floating IP list to include port forwarding information
+    for fip in floating_ips:
+        # no need to do this for floating IPs associated with a port,
+        # as port forwarding is irrelevant in such a case
+        if not fip.port_id:
+            pfws = list(connection.network.port_forwardings(fip))
+            if len(pfws):
+                fip.port_id = pfws[0].internal_port_id
+                pfw_ports = ["%s:%s" % (pfw.internal_port,
+                                        pfw.external_port)
+                             for pfw in pfws]
+                fip.floating_ip_address = "%s (%s)" % (
+                    fip.floating_ip_address, ','.join(pfw_ports))
+
+    data = []
+    for port in ports:
+        if not filter_node:
+            node_name = next((node for node in nodes
+                              if node.id == port.node_id), None).name
+
+        neutron_port_id = port.internal_info.get('tenant_vif_port_id')
+        neutron_port = None
+
+        if neutron_port_id:
+            neutron_port = next((np for np in neutron_ports
+                                 if np.id == neutron_port_id), None)
+
+        if neutron_port is not None:
+            network_id = neutron_port.network_id
+
+            if not filter_network or filter_network.id == network_id:
+                network_names, _, fixed_ips \
+                    = utils.get_full_network_info_from_port(
+                        neutron_port, connection, networks_dict)
+                floating_ip_addresses, floating_network_names \
+                    = utils.get_floating_ip(neutron_port_id,
+                                            floating_ips,
+                                            networks_dict)
+                data.append([node_name, port.address,
+                             neutron_port.name,
+                             "\n".join(network_names),
+                             "\n".join(fixed_ips),
+                             "\n".join(floating_network_names)
+                             if floating_network_names else None,
+                             "\n".join(floating_ip_addresses)
+                             if floating_ip_addresses else None]
+                            )
+        elif not filter_network:
+            data.append([node_name, port.address,
+                         None, None, None, None, None])
+
+    return ["Node", "MAC Address", "Port", "Network", "Fixed IP",
+            "Floating Network", "Floating IP"], data
```

Comment:
I'll take a closer look tomorrow, but one thing sticks out: we want to return `data`, and not the column headers. I'd consider the headers part of the formatting, and that belongs in `python-esiclient`. Another way to think of this, is if you want to use this function in the UI, you're only going to want `data`.

Reply:
Taking that one step forward, I wonder if this might be best returned as a list of dictionaries, where each dictionary is something like:
```
{
    "node": <node object>,
     "ports": [{
         "baremetal_port": <baremetal port object>,
         "neutron_port": <neutron port object>,
         "floating_ip": <floating ip information>
       },
       <additional ports>
}
```

Then the client gets to interpret the information however they'd like.